### PR TITLE
CM-700: Add custom validation to sites

### PR DESCRIPTION
**Pull request by ayumi-oxd**
### Jira Ticket:
CM-700

### Jira Ticket URL:
https://bcparksdigital.atlassian.net/browse/CM-700

### Description:
- Add custom validation to the sites/park-sub-pages to check if it has a relation with the protectedAreas
- The error message will be displayed with these conditions:
  - When an editor creates a new entry and tries to save it without a relation
  - When an editor removes a relation and tries to save an entry

### CODE REVIEW

File: src/cms/src/api/site/content-types/site/lifecycles.js [hunk 1]
```diff
@@ -1,25 +1,44 @@
...
-        validator.slugNoTrailingDashValidator(data.slug)
-    },
-    beforeUpdate(event) {
-        const { data, where, select, populate } = event.params;
-        validator.slugCharacterValidator(data.slug)
-        validator.slugNoLeadingSlashValidator(data.slug)
-        validator.slugNoLeadingDashValidator(data.slug)
-        validator.slugNoTrailingDashValidator(data.slug)
-    }
+  async beforeCreate(event) {
```
Comment (molund) [inline src/cms/src/api/site/content-types/site/lifecycles.js:28]: `where.id` won't have a value yet on beforeCreate

File: src/cms/src/helpers/validator.js [hunk 2]
```diff
@@ -2,6 +2,18 @@ const utils = require('@strapi/utils');
...
 module.exports = {
+  // saving without a protectedArea relation is not allowed
+  protectedAreaConnectValidator: function (protectedArea) {
+    if (protectedArea.connect.length === 0) {
+      throw new ApplicationError('Please add protectedArea relation.');
+    }
+  },
+  // removing a protectedArea relation is not allowed
+  protectedAreaDisconnectValidator: function (protectedArea) {
+    if (protectedArea.disconnect.length > 0) {
```
Comment (molund) [inline src/cms/src/helpers/validator.js:13]: This doesn't work right when you change to a different protectedArea.  Maybe try something like this instead?
```
    if (protectedArea.disconnect.length > 0 && protectedArea.connect.length === 0) {
```

File: src/cms/src/api/site/content-types/site/schema.json [hunk 3]
```diff
@@ -58,7 +58,8 @@
       "type": "relation",
       "relation": "manyToOne",
       "target": "api::protected-area.protected-area",
-      "inversedBy": "sites"
+      "inversedBy": "sites",
+      "required": true
```
Comment (molund) [inline src/cms/src/api/site/content-types/site/schema.json:62]: Does this `required` attribute actually do anything?  I noticed you don't have one for park-sub-pages

File: src/cms/src/helpers/validator.js (hunk 2 above)
Comment (ayumi-oxd) [reply src/cms/src/helpers/validator.js]: That works, thanks!

File: src/cms/src/api/site/content-types/site/schema.json (hunk 3 above)
Comment (ayumi-oxd) [reply src/cms/src/api/site/content-types/site/schema.json:62]: I have `required` for `site` and `park-sub-page` 🙂 
It just displays `*` on Strapi UI to indicate it's a required field.

Screenshot 2023-05-30 at 9 35 20 AM

**Review by molund — APPROVED** [review]

**Pull request merged**